smpclient/__init__.py

```diff
@@ -50,10 +56,18 @@ async def upload(
     ) -> AsyncIterator[int]:
         """Iteratively upload an `image` to `slot`, yielding the offset."""
 
+        max_packet_size = self._transport.max_unencoded_size
+        if isinstance(self._transport, SMPBLETransport):
+            mcumgr_parameters = await asyncio.wait_for(self.request(MCUMgrParametersRead()), timeout=1.0)
+            if success(mcumgr_parameters):
+                max_packet_size = mcumgr_parameters.buf_size
```

Comment:
I like the idea of using this!  I think it could be used for all of the transports, not just BLE.  The idea would be:

* User provides MTU -> don't make this request, use user's value
* User doesn't provide MTU -> make request and error if it fails

I ran this request to an SMP server running in MCUBoot and it returned the not supported RC.  I don't see an (easy) way to enable it.  So it's definitely necessary to allow the user to provide MTU in some cases.

https://github.com/nrfconnect/sdk-mcuboot/blob/main/boot/zephyr/Kconfig.serial_recovery

Reply:
I'm not sure but it seems to be valid only for ble on zephyr side. I would have to double check though.

Reply:
those do the trick
```
MCUMGR_GRP_OS
MCUMGR_GRP_OS_MCUMGR_PARAMS
NCS_SAMPLE_MCUMGR_BT_OTA_DFU_SPEEDUP
```

Reply:
Reference from the NCS speedup Kconfig:

```
config NCS_SAMPLE_MCUMGR_BT_OTA_DFU_SPEEDUP
	bool "MCUmgr OTA DFU speedup"
	depends on BT_CTLR || (NCS_SAMPLE_HCI_RPMSG_CHILD_IMAGE || NCS_SAMPLE_MULTIPROTOCOL_RPMSG_CHILD_IMAGE)
	help
	  Enable this option to speed up the OTA DFU transfer over Bluetooth.
	  This option extends the Bluetooth buffers to extend Bluetooth MTU
	  size. The buffer configuration should be the same in the Network Core
	  and Application Core image.

	  This configuration option is automatically applied for the Network
	  Core child image if the NCS_SAMPLE_MCUMGR_BT_OTA_DFU_SPEEDUP option is
	  enabled in the main image.

if NCS_SAMPLE_MCUMGR_BT_OTA_DFU_SPEEDUP

config BT_L2CAP_TX_MTU
	default 498 if BT_HCI_HOST

config BT_BUF_ACL_TX_SIZE
	default 251

config BT_BUF_ACL_RX_SIZE
	default 502

config BT_CTLR_DATA_LENGTH_MAX
	default 251 if BT_CTLR

endif # NCS_SAMPLE_MCUMGR_BT_OTA_DFU_SPEEDUP
```
One of the apps I'm testing with has these options which I got from some Nordic forum thread:
```
# Allow for large Bluetooth data packets.
CONFIG_BT_L2CAP_TX_MTU=498
CONFIG_BT_BUF_ACL_RX_SIZE=502
CONFIG_BT_BUF_ACL_TX_SIZE=502
CONFIG_BT_CTLR_DATA_LENGTH_MAX=251
CONFIG_HEAP_MEM_POOL_SIZE=4096

# Enable the mcumgr Packet Reassembly feature over Bluetooth and its configuration dependencies.
# MCUmgr buffer size is optimized to fit one SMP packet divided into five Bluetooth Write Commands,
# transmitted with the maximum possible MTU value: 498 bytes.
CONFIG_MCUMGR_TRANSPORT_BT_REASSEMBLY=y
CONFIG_MCUMGR_TRANSPORT_NETBUF_SIZE=2475
CONFIG_MCUMGR_GRP_OS_MCUMGR_PARAMS=y
CONFIG_MCUMGR_TRANSPORT_WORKQUEUE_STACK_SIZE=4608
```

Reply:
I'm adding some test scripts for this and here is the return that I get:

```
buf_size=2475 buf_count=4
```
It is because the overlay that the docs suggest for the BLE OTA DFU sample set netbuf to that size:
`zephyr\samples\subsys\mgmt\mcumgr\smp_svr\overlay-bt.conf`

```
# Enable the mcumgr Packet Reassembly feature over Bluetooth and its configuration dependencies.
# MCUmgr buffer size is optimized to fit one SMP packet divided into five Bluetooth Write Commands,
# transmitted with the maximum possible MTU value: 498 bytes.
CONFIG_MCUMGR_TRANSPORT_BT_REASSEMBLY=y
CONFIG_MCUMGR_TRANSPORT_NETBUF_SIZE=2475
CONFIG_MCUMGR_GRP_OS_MCUMGR_PARAMS=y
CONFIG_MCUMGR_TRANSPORT_WORKQUEUE_STACK_SIZE=4608
```
Windows is telling me that the BLE MTU is 20B...